src/common/utils/genericsubscriptionmanager.hpp

```diff
@@ -0,0 +1,275 @@
+#include <future>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_set>
+
+#include <grpcpp/grpcpp.h>
+
+#include <common/logger/logmodule.hpp>
+#include <core/common/tools/error.hpp>
+
+namespace aos::common::utils {
+
+/**
+ * Generic subscription manager template that handles gRPC stream subscriptions.
+ *
+ * @tparam TStub gRPC stub type (e.g., IAMPublicCertService::Stub).
+ * @tparam TListener Listener interface type (e.g., CertListenerItf).
+ * @tparam TProtoMsg Protobuf message type (e.g., iamanager::v6::CertInfo).
+ * @tparam TAosType AOS native type (e.g., CertInfo).
+ * @tparam TRequest Protobuf request type (e.g., SubscribeCertChangedRequest).
+ */
+template <typename TStub, typename TListener, typename TProtoMsg, typename TAosType, typename TRequest>
+class GenericSubscriptionManager {
+public:
+    // Type aliases for function pointers
+    using ReaderPtr   = std::unique_ptr<grpc::ClientReader<TProtoMsg>>;
+    using ReaderFunc  = ReaderPtr (TStub::*)(grpc::ClientContext*, const TRequest&);
+    using ConvertFunc = std::function<Error(const TProtoMsg&, TAosType&)>;
+    using NotifyFunc  = std::function<void(TListener&, const TAosType&)>;
+
+    /**
+     * Constructor.
+     *
+     * @param stub gRPC service stub.
+     * @param request Subscription request.
+     * @param readerFunc Pointer to stub's subscription method.
+     * @param convertFunc Function to convert proto message to AOS type.
+     * @param notifyFunc Pointer to listener's notification method.
+     * @param logContext Context string for logging.
+     */
+    GenericSubscriptionManager(TStub* stub, TRequest request, ReaderFunc readerFunc, ConvertFunc convertFunc,
+        NotifyFunc notifyFunc, std::string logContext = "")
+        : mStub(stub)
+        , mRequest(std::move(request))
+        , mReaderFunc(readerFunc)
+        , mConvertFunc(std::move(convertFunc))
+        , mNotifyFunc(notifyFunc)
+        , mLogContext(std::move(logContext))
+    {
+    }
+
+    /**
+     * Destructor.
+     */
+    ~GenericSubscriptionManager() { Close(); }
+
+    // Non-copyable, non-movable
+    GenericSubscriptionManager(const GenericSubscriptionManager&)            = delete;
+    GenericSubscriptionManager& operator=(const GenericSubscriptionManager&) = delete;
+    GenericSubscriptionManager(GenericSubscriptionManager&&)                 = delete;
+    GenericSubscriptionManager& operator=(GenericSubscriptionManager&&)      = delete;
+
+    /**
+     * Adds a subscriber.
+     *
+     * @param listener Listener to add.
+     * @return Error.
+     */
+    Error AddSubscriber(TListener& listener)
+    {
+        std::lock_guard lock {mMutex};
+
+        LOG_INF() << "Add subscriber" << Log::Field("context", mLogContext.c_str());
+
+        if (!mSubscribers.insert(&listener).second) {
+            return Error(ErrorEnum::eAlreadyExist, "subscriber already exists");
+        }
+
+        if (mSubscribers.size() == 1 && !mRunning) {
```

Comment:
Looks like you don't need `mRunning` flag. You can check same with `mSubscribers.size()`